### PR TITLE
Create DiscussionLocks

### PR DESCRIPTION
**Pull request by atsmith813**
## What type of PR is this? (check all applicable)
- [x] Feature

## Description
This PR introduces functionality to allow **article authors, admins, and super admins** to lock discussions which basically prevents the ability for users to add any further comments to an article. We've also added some logic to allow a permitted user to include a `reason` that will display publicly when a discussion is locked and `notes` which are only visible to users permitted to lock the discussion.

## Related Tickets & Documents
closes https://github.com/forem/forem/issues/9619

## QA Instructions, Screenshots, Recordings
1. Pull down this code.
2. Fire up the Rails server locally - `bin/startup`.
3. Login.
4. Write a post.
5. Go to `/manage` on your published post.
6. Click `Lock discussion`.
7. Add a `reason` and `notes`.
8. Lock the discussion.
9. View your article - you should see the discussion is locked and you're unable to add a comment.
10. Go back to `/manage` on your post.
11. Unlock the discussion.
12. Go back to your post and leave a comment.

**Lock discussion**
lock_discussion

**Confirm lock discussion**
lock_discussion_confirm

**Lock discussion success**
success

**Post with discussion locked (with reason)**
with_reason

**Post with discussion locked (without reason)**
without_reason

**Unlock discussion**
unlock_discussion

**Confirm unlock discussion**
unlock_discussion_confirm

**Error example**
error

## Added tests?
- [x] Yes

## [Forem core team only] How will this change be communicated?
- [x] I'm not sure how best to communicate this change and need help

wedding_crashers_lock_it_up_gif

### CODE REVIEW

**Comment by citizen428** [comment]
This is probably outside the scope of this PR but I wonder if we want to add some `ResponseTemplate`s for this at one point. I think it could make sense as there are probably some recurring reasons (too many offtopic comments, discussion became uncivilized, etc.) and standardizing the language could be helpful.

**Review by aitchiss — CHANGES_REQUESTED** [review]
This is looking great! I've added a couple of comments with some small suggestions and a change request, but all very minor. 

A couple of other things I noticed:

- The (un)lock confirm page doesn't have a page title (as in the title that appears in the browser tab) - can we add one?

- Should we hide the comment reply button when the thread is locked? At the moment you can draft a comment and hit submit to get an error message:

https://user-images.githubusercontent.com/20773163/120619518-f4017d00-c453-11eb-8e31-a4f2f5c8a197.mp4

- I know this will be unpopular... but I think we should also add a Cypress test for this. It should be straightforward now since there's no funky async javascript behind the scenes, but let me know what you think.

File: app/models/discussion_lock.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+class DiscussionLock < ApplicationRecord
+  belongs_to :article
+  belongs_to :locking_user, class_name: "User"
+
+  before_validation :nullify_notes_and_reason
```
[the remark this answers is not in the document]
Comment (atsmith813) [reply app/models/discussion_lock.rb:5]: Updated.

File: app/policies/discussion_lock_policy.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
+class DiscussionLockPolicy < ApplicationPolicy
+  def create?
+    authorized_user?
```
[the remark this answers is not in the document]
Comment (atsmith813) [reply app/policies/discussion_lock_policy.rb]: Updated.

File: app/policies/discussion_lock_policy.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  end
+
+  def destroy?
+    create?
+  end
+
+  def permitted_attributes
+    %i[article_id notes reason]
```
[the remark this answers is not in the document]
Comment (atsmith813) [reply app/policies/discussion_lock_policy.rb]: Updated.

File: app/controllers/discussion_locks_controller.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    if @discussion_lock.save
+      bust_article_cache(article)
+
+      flash[:success] = "Discussion was successfully locked!"
+    else
+      flash[:error] = "Error: #{@discussion_lock.errors_as_sentence}"
+    end
+
+    redirect_to "#{article.path}/manage"
```
[the remark this answers is not in the document]
Comment (atsmith813) [reply app/controllers/discussion_locks_controller.rb:20]: I added it to the routes, but using it in the controller is more work.

`article.path` is a field on `Article` with the `path` already set as a string. To use this named route helper, we'd need to include the `User` record or rely on `cached_user.username` which could be stale.

File: app/controllers/articles_controller.rb [hunk 5]
```diff
@@ -195,6 +196,24 @@ def admin_unpublish
     end
   end
 
+  def discussion_lock_confirm
```
[the remark this answers is not in the document]
Comment (citizen428) [reply app/controllers/articles_controller.rb:199]: > Since the articles controller is "too far gone" at this point 

The [broken windows theory of software development](https://medium.com/@matryer/broken-windows-theory-why-code-quality-and-simplistic-design-are-non-negotiable-e37f8ce23dab) strikes again.

Not criticizing your decision here, I think your argument is valid and reasonable. But it also serves as a warning as to why we want to ensure new code doesn't get into a state like this.

File: app/controllers/discussion_locks_controller.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (citizen428) [reply app/controllers/discussion_locks_controller.rb:20]: Ugh, I see. This was a complexity I overlooked.

File: app/views/articles/manage.html.erb [hunk 6]
```diff
@@ -5,11 +5,11 @@
...
       </div>
-      <% if flash[:pins_error] %>
-        <h3 class="manage-page-error">Uh Oh! <%= flash[:pins_error] %> 😬</h3>
+      <% if flash[:error] %>
+        <h3 class="manage-page-error">Uh Oh! <%= flash[:error] %> 😬</h3>
       <% end %>
-      <% if flash[:pins_success] %>
-        <h3 class="manage-page-success"><%= flash[:pins_success] %></h3>
+      <% if flash[:success] %>
+        <h3 class="manage-page-success"><%= flash[:success] %></h3>
```
[the remark this answers is not in the document]
Comment (atsmith813) [reply app/views/articles/manage.html.erb:12]: Updated.

**Comment by atsmith813** [comment]
@aitchiss I can add some Cypress tests for this. Hiding that reply button is tricky as well because we render that markup in JS _and_ server-side. I'll have to find a way to communicate with the JS that the article was locked, bringing us back to a similar situation as the last PR 😭 

I'll work on that 👍🏽

**Comment by atsmith813** [comment]
Hid **reply** - need to work on cypress tests next 💯

File: db/migrate/20210531152321_create_discussion_locks.rb [hunk 7]
```diff
@@ -0,0 +1,12 @@
+class CreateDiscussionLocks < ActiveRecord::Migration[6.1]
+  def change
+    create_table :discussion_locks do |t|
+      t.references :article, null: false, foreign_key: true, index: { unique: true }
+      t.references :locking_user, references: :users, foreign_key: { to_table: :users }, null: false
```
Comment (jgaskins) [inline db/migrate/20210531152321_create_discussion_locks.rb:5]: Nice work on adding FKs and constraints/indexes! :100:

File: app/models/discussion_lock.rb [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+  validates :article_id, presence: true, uniqueness: true
+  validates :locking_user_id, presence: true
+
+  private
+
+  def nullify_blank_notes_and_reason
+    # Prevent blank strings from beings saved to the DB
+    self.notes = nil if notes.blank?
+    self.reason = nil if reason.blank?
```
Comment (jgaskins) [inline app/models/discussion_lock.rb:15]: What's the reason behind this choice?

To be clear, I agree very much in principle with the idea of not having both `NULL` and empty strings. I just haven't noticed us using this particular idea elsewhere.

File: app/models/user.rb [hunk 9]
```diff
@@ -186,6 +186,7 @@ class User < ApplicationRecord
   has_many :comments, dependent: :destroy
   has_many :created_podcasts, class_name: "Podcast", foreign_key: :creator_id, inverse_of: :creator, dependent: :nullify
   has_many :credits, dependent: :destroy
+  has_many :discussion_locks, dependent: :destroy, inverse_of: :locking_user, foreign_key: :locking_user_id
```
Comment (jgaskins) [inline app/models/user.rb:189]: Is it even possible to destroy the user if they still have discussion locks active due to the FK constraint? If ActiveRecord doesn't apply a `DEFERRED` clause to the FK constraint, this line may not have any effect. Should we set the FK to `ON DELETE CASCADE` instead?

That brings up a product-related question, actually. Does it make sense to silently unlock discussions simply because the user that locked them was deleted? Or would it make sense instead to have some kind of "ghost" user like GitHub does, where whenever a user is deleted, all their comments, PRs, etc, all get assigned to the ghost user?

File: app/models/discussion_lock.rb (hunk 8 above)
Comment (atsmith813) [reply app/models/discussion_lock.rb:15]: It's because the `notes` and `reason` fields are optional and if a user doesn't type anything (we expect this to be the most common case) then empty strings (`""`) are passed through and saved.

I don't feel strongly about this - I'm happy to just remove it if folks feel some type of way about this.

File: app/models/user.rb (hunk 9 above)
Comment (jgaskins) [reply app/models/user.rb:189]: Right, but I mean, will Postgres even _allow_ us to delete that user with the FK constraint in place?

Comment (jgaskins) [reply app/models/user.rb:189]: Huh, TIL ActiveRecord destroys the associated records _before_ destroying the parent record. I thought the `dependent: :destroy` stuff was run with `after_destroy` hooks, but apparently it's run with the `before_destroy`. 💡 

```
  DiscussionLock Destroy (0.2ms)  DELETE FROM "discussion_locks" WHERE "discussion_locks"."id" = $1  [["id", 5]]
 <snip>
  User Destroy (7.9ms)  DELETE FROM "users" WHERE "users"."id" = $1  [["id", 9]]
```

**Review by rhymes — COMMENTED** [review]
Nice! I left some notes.

Haven't tested it yet as I'm the middle of adding 10k articles on my local DB 🤣 
Will do as soon as I finish!

File: app/models/discussion_lock.rb (hunk 8 above)
Comment (citizen428) [reply app/models/discussion_lock.rb:15]: I know that you're not the biggest fan of metaprogramming @jgaskins but we could really easily share this behavior via a module builder. Here's the quick and dirty 5-minute version:

```ruby
module AttributeCleaner
  def self.for(*attrs)
    Module.new do
      def self.included(klass)
        klass.before_validation :nullify_blank_attributes
      end

      define_method :nullify_blank_attributes do
        attrs.each { |attr| public_send("#{attr}=", nil) if public_send(attr).blank? }
      end
    end
  end
end
```

You then include it like `include AttributeCleaner.for(:attr1, :attr2).

Here it is in action (in a Rails console where I loaded the code above):

```ruby
Profile.include(AttributeCleaner.for(:location))
p = Profile.new(location: "    ")
p.valid? 
#=> false
p.location
#=> nil
```

File: cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js [hunk 10]
```diff
@@ -0,0 +1,107 @@
+describe('Lock discussion', () => {
```
Comment (aitchiss) [inline cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js:1]: it's tiny thing but there's a typo in the subdirectory - `mangageFlows` -> `manageFlows`

File: cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    cy.findByRole('link', {
+      name: /Lock discussion/,
+    });
+
+  const getDiscussionUnlockButton = () =>
+    cy.findByRole('link', {
+      name: /Unlock discussion/,
+    });
+
+  const getDiscussionLockSubmitButton = () => cy.get('.crayons-btn--danger');
```
Comment (aitchiss) [inline cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js:12]:
```suggestion
  const getDiscussionLockSubmitButton = () => cy.findByRole('button', { name: 'Lock discussion' });
```

Just a suggestion, but changing this to another `findByRole` would help us test closer to the user experience (since a user finds by the button name rather than classname)

(same goes for the other test file that uses the same)

File: cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+      getDiscussionLockButton().click();
+
+      cy.findByRole('heading', {
+        name: 'Are you sure you want to lock the discussion on this article?',
+      }).should('exist');
+
+      cy.findByRole('button', {
+        name: /Lock discussion/,
+      }).should('exist');
+    });
```
Comment (aitchiss) [inline cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js]: I don't have any strong feeling about it, but I think this _could_ be combined with the previous test

File: cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+
+    it('should allow a user to supply a reason and notes', () => {
+      getDiscussionLockButton().click();
+      cy.get('input[name="discussion_lock[notes]"]').should('exist');
+      cy.get('input[name="discussion_lock[reason]"]').should('exist');
```
Comment (aitchiss) [inline cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js:58]:
```suggestion
cy.findByRole('textbox', {
        name: 'Reason for locking the discussion (optional) - this will be publicly displayed',
      }).should('exist');
      cy.findByRole('textbox', {
        name: 'Notes (optional) - this is only visible to you and admins',
      }).should('exist');
```

I think we should change this, so it's more focused on the user-facing label

File: cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+            getDiscussionLockSubmitButton().click();
+            cy.visit(articlePath);
+          });
+        });
+      });
+    });
+
+    it('should show the discussion lock on an article', () => {
+      cy.get('#discussion-lock').should('exist');
+    });
```
Comment (aitchiss) [inline cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js:93]: I don't think we need this as a test, as the other two tests here make sure we're showing the user what we should

File: cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+      cy.get('#discussion-lock').should(($discussionLock) => {
+        expect($discussionLock).to.contain.text(exampleReason);
+      });
+    });
+
+    it('should not show the discussion lock notes on an article', () => {
+      cy.get('#discussion-lock').should(($discussionLock) => {
+        expect($discussionLock).not.to.contain.text(exampleNotes);
+      });
+    });
```
Comment (aitchiss) [inline cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js:105]: Just a thought - could these verification steps be tacked on the end of the previous tests e.g. "should allow a user to lock a discussion"? I'm just thinking when you lock the discussion, you're on the "manage" page and could then just click the article link and check the notice appears?

It would save the additional test setup for these ones, and it would mean the tests cover the full flow from being able to submit the form to seeing the correct UI on the article page.

No worries if there's a reason that doesn't work though.

File: cypress/integration/articleFlows/mangageFlows/unlockDiscussion.spec.js [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+        });
+      });
+    });
+
+    it('should allow a user to unlock a discussion', () => {
+      getDiscussionUnlockButton().click();
+      getDiscussionLockSubmitButton().click();
+
+      getDiscussionLockButton().should('exist');
+    });
```
Comment (aitchiss) [inline cypress/integration/articleFlows/mangageFlows/unlockDiscussion.spec.js:40]: Related to my comment above, but if possible I think it would be great if we could then visit the article page and check that the comment form appears as expected, just to test the desired effect came from the form submission

**Review by aitchiss — COMMENTED** [review]
I added some thoughts/suggestions on the cypress tests 🙂

File: cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js (hunk 11 above)
Comment (atsmith813) [reply cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js]: I originally did this intentionally so these specs don't fail for a quick "text change" on the button. I was thinking we don't mind what the button text is as much as that there is a button there that does what we expect, in general.

Regardless, I implemented your suggestion.

File: cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js (hunk 13 above)
Comment (atsmith813) [reply cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js]: Similar to - https://github.com/forem/forem/pull/13905#discussion_r647787780, I originally didn't think the text mattered as much.

Regardless, I implemented your suggestions.

File: cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js (hunk 15 above)
Comment (atsmith813) [reply cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js]: I didn't do that for 2 reasons:
1. Managing the state to get the article path is annoying. The before actions here are different than the ones above.
2. The _displaying_ of discussion locks and the _creation/deletion_, in my mind, are 2 totally different things. The code for each doesn't touch the other. I figured this "degree of separation" in the test would also reflect that. Technically, we could create discussion locks and temporarily disable displaying them for some reason. In that edge case, these tests would be "easier" to manage - skip the second set.

If we really feel strongly about it, we can smash it all into one.

File: app/policies/discussion_lock_policy.rb [hunk 17]
```diff
@@ -0,0 +1,21 @@
+class DiscussionLockPolicy < ApplicationPolicy
+  PERMITTED_ATTRIBUTES = %i[article_id notes reason].freeze
+
+  def create?
+    user_author? || minimal_admin?
```
[the remark this answers is not in the document]
Comment (citizen428) [reply app/policies/discussion_lock_policy.rb]: Hmm, I think this is a bit debatable. Even if a user got suspended they still own their content. It seems a bit off that people would be able to comment on the article but the original author has no way to respond. If anything I'd go as far as automatically locking the posts of suspended users and disallowing them from unlocking.

File: cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js (hunk 15 above)
Comment (aitchiss) [reply cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js]: I don't feel strongly about it at all 😄  That all makes sense!

I was thinking more from the user POV, after you lock a discussion you would naturally want to see some change reflected on the article page so it would be one user flow (and by the time you get to the "Manage" page you have a link to the article you can click, so we don't need to worry about the article path state any more). 

But I'm happy for them to be separate, it was just a thought!

File: cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js (hunk 13 above)
Comment (aitchiss) [reply cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js]: I think the text matters - e2e tests are really to test from a user's point of view. If we accidentally push a change that unlinks the label from the input, these tests will catch that if we use the input name

File: cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js (hunk 11 above)
Comment (aitchiss) [reply cypress/integration/articleFlows/mangageFlows/lockDiscussion.spec.js]: I commented on the other similar one below, but the general philosophy with e2e tests is to test as close to how a user would use the app as possible. With that in mind, it makes more sense to identify buttons/links/etc by their names, rather than CSS class or ID attribute because those things aren't exposed to users (and equally could change in the future).

It also helps catch accessibility issues on a page, as `findByRole` (unlike `get`) will only succeed if the button is visible to both the UI and assistive technology - e.g. if I add a `hidden` class to the button, the `get` selector will still return the element whereas `findByRole` will fail. Which is pretty neat!

Just mentioning all this for context 🙂

**Review by aitchiss — APPROVED** [review]
Tested locally and this is looking good to me 🚀  Thanks for making those Cypress changes 🙂

**Review by rhymes — APPROVED** [review]
Works well, good job @atsmith813!

I have a "product" question: I tried to lock a discussion after leaving a comment. When the discussion is locked, I can still edit or delete the comment. 

Is this okay? I don't know the right answer, just asking because it felt a bit odd to be able to change the text of a comment if a discussion is locked. I'm thinking of people altering their own comments after a discussion has been locked for gaslighting or abuse. Maybe we shouldn't allow them to be able to do that.

I wonder if we should ping Product or Community to have a definite answer for this, as AFAIK we don't have a RFC for it
